### ISSUE TITLE
Document running transformations on GPU

### ISSUE THREAD

**Issue by Borda** [Lightning-AI/lightning-flash#1186]
## 🚀 Feature

I am curious if we could integrate / run data loader transforms on GPU

### Motivation

it may even speed up the training process

### Alternatives

see: https://github.com/PhoenixDL/rising

### Additional context

@justusschock

**Comment by krshrimali** [comment]
Interesting. Just putting some of my thoughts here for discussion:

Applying transforms on GPU or loading data will also utilize some GPU memory, right? Most of the datasets in DL are usually in GBs, and I suspect that a lot of OOM errors might occur. Unless and until we only load a batch of data into memory, apply transforms, and then transfer back to the CPU.

From the README of `rising` repo:

> Running augmentations on the GPU is only efficient if they can be executed in a batched fashion to maximize the parallelization GPUs can provide

I didn't see any benchmarks on the repo, but I can try on my system to see if there is any improved time and if it goes OOM.

**Comment by ethanwharris** [comment]
So, we do have some support here. The InputTransform has the hook `per_batch_transform_on_device` which can be used to provide transforms that execute per batch on the GPU.

There is a small example here that shows the general idea: https://github.com/PyTorchLightning/lightning-flash#flash-transforms

But it would be nice to document this better and make sure it is tested and has a smooth integration with rising 😃

**Comment by justusschock** [comment]
@krshrimali For medical images the pipeline usually is:

1. fetch dataset statistics (may require to load every sample once, important: don't cache it)
2. determine which transforms to use based on 1
3. load a single batch of images (can be 1-N samples)
4. augment them (can be very costly on CPU, especially resampling 3D images is very compute intensive)
5. transfer batch to GPU
6. do training with current batch

where the steps 3-6 are repeated for every batch in the dataset. This means that for most use-cases, the whole dataset is never loaded completely but only single batches.

Applying augmentations on the GPU is most efficient using batches rather than single samples whenever possible and can't easily be done inside the dataset (due to overheads with the cuda context, which would be created by every single dataloader worker if they were accessing the GPU).

With rising, the pipeline is 

1. fetch dataset statistics (may require to load every sample once, important: don't cache it)
2. determine which transforms to use based on 1
3. load a single batch of images (can be 1-N samples)
4. transfer batch to GPU 
6. augment them (should be less costly than on CPU, can usually be done without gradient computation/AD graph building and with proper memory management shouldn't result in an overhead once augmentation is finished, since the original batch ["pre augmentation"] doesn't have to be stored; An exception of this is of course test-time augmentation.)
7. do training with current batch

With those principles in mind, we designed rising and those principles were regarded, when designing the original data processing pipeline in flash (by @tchaton and me) which was then revised multiple times by @ethanwharris .


@ethanwharris In the past, there was also a possibility to apply per-sample transforms on device. Has this been removed? Because transformations like resizing (also costly for a 3d image) have to be applied per-sample if the samples are of different sizes and thus sometimes has to be done before even batching them :)

**Comment by ethanwharris** [comment]
Hey @justusschock we still have the `per_sample_transform_on_device` although I would be interested to see some concrete use-cases / examples for it. That way we can start to benchmark as it's a source of a lot of complexity in the code and it might be that the boost you get from resizing on device is offset by not having each sample transformed in parrallel.

In general it would be nice if we had like an `accelerating transforms` guide that desicussed on_device transforms in detail and gave some concrete examples along with speedup. That would allow us to better test / support this functionality going forward.
